### PR TITLE
Airflow tags

### PR DESCRIPTION
**Pull request by kik-kik**
Airflow tags proposal and discussions:
https://docs.google.com/document/d/1LqOCmadsC6kPTusqyczHha-mdG_DVsNWK4-a1-NHI_8/edit#

### CODE REVIEW

**Review by jklukas — CHANGES_REQUESTED** [review]
I'm not convinced of the need for a tags.yaml file. I'd prefer that we verbosely pass tags into the `DAG` constructor. Is there a reason to avoid that?

**Comment by kik-kik** [comment]
I just figured it would be "easier" to have an external config file for pulling the dags, plus additionally as part of that process there is some validation happening. Of course, we could just have a function call when setting the tags which could validate them. Thinking something like this:

```python

tags = [list of tags]

with DAG(
    "my_dag",
    [...],
    tags=validate_tags(tags)
) as dag:
[...]
```

What do you think about that?

**Comment by jklukas** [comment]
> we could just have a function call when setting the tags which could validate them

We'd still have the case of users being able to create a new DAG and not remember to include the `validate_tags` function.

I'd like to see us have a CI task that's responsible for validating tags rather than needing to have code within the DAGs to do it. The `test-environment` CI task makes sure that each DAG is able to load, but uses the CLI. Would it be possible to adapt that task to handle validating tags?

**Comment by kik-kik** [comment]
so I think one option here could also be to extend the standard `DAG` object and make tags field a required field and do verification at this stage? What do you think @jklukas ?

**Comment by jklukas** [comment]
> so I think one option here could also be to extend the standard `DAG` object and make tags field a required field and do verification at this stage?

That's worth exploring. My first thought is that I don't see a way of enforcing that new DAGs use our custom subclass.

I'd love to hear opinions from @scholtzan who was involved in reviewing the tags proposal. Do you have any opinions about the options here for validating tags?

**Comment by kik-kik** [comment]
@jklukas I also changed how I set my tags, decided to have main file with enums defined (valid options) and use those in DAGs themselves. Please let me know what you think about that. This way at least we can easily modify the value for each tier if we choose to do so with a change only required in a single file (+ test file) instead of each DAG. Plus it gives you autocompletion when working with the DAG.

**Comment by jklukas** [comment]
> decided to have main file with enums defined (valid options) and use those in DAGs themselves. Please let me know what you think about that

This feels like a big improvement. It enables validation of values and ends up being basically self-documenting. If you see a DAG using these enums to set tags, then it naturally leads you to looking up the tags module to discover what the tag options are.

**Comment by kik-kik** [comment]
Awesome @jklukas. I'm glad to hear you seem to like this approach. I'm not sure I'm clever enough to come up with anything more sophisticated... 😅

**Review by scholtzan — APPROVED** [review]
Looks good to me. I think we might need to make some changes to the scripts copying the generated DAGs over to Airflow to also include `utils/tags.py`. @haroldwoo might have more context on where that change needs to be made

**Comment by jklukas** [comment]
> @haroldwoo might have more context on where that change needs to be made

I do think we need to be careful about how this is rolled out. We have historically had issues with DAGs that depend on code in other modules. Even if everything is defined inside this repo, it's possible to have skew as to what code is available to workers vs. the web interface. We have had DAGs fail due to depending on module code that isn't successfully updated. This sometimes requires cycling workers.

I do think we should check with @haroldwoo, and this may fall under something that needs to wait until after the infrastructure change freeze.

**Review by jklukas — APPROVED** [review]
r+wc

I'm happy with the changes here. I've updated the title with "DO NOT MERGE" thought pending more info on how Airflow deployment works and whether this requires coordination with SRE.

**Comment by kik-kik** [comment]
@jklukas should we perhaps bring it up on Thursday during the `Data Platform Infrastructure Working Group`?

**Comment by jklukas** [comment]
> should we perhaps bring it up on Thursday during the Data Platform Infrastructure Working Group, I believe normally someone from Data SRE is there?

Yes, we can do that, although we may end up needing to redirect to hwoo anyway. You could also file a DSRE jira issue and tag hwoo there to ask if he knows how this would need to be rolled out.

**Comment by kik-kik** [comment]
> > should we perhaps bring it up on Thursday during the Data Platform Infrastructure Working Group, I believe normally someone from Data SRE is there?
> 
> Yes, we can do that, although we may end up needing to redirect to hwoo anyway. You could also file a DSRE jira issue and tag hwoo there to ask if he knows how this would need to be rolled out.

okay, cool. I created a JIRA ticket with SRE and tagged hwoo there: https://mozilla-hub.atlassian.net/browse/DSRE-379

File: requirements.in
Comment (haroldwoo) [inline requirements.in:22]: im surprised we didnt run pytest in our circleci config. Maybe it was removed at some point

Comment (haroldwoo) [reply requirements.in:22]: can you add the running of pytests to our circleci config if it passes?

**Review by haroldwoo — COMMENTED** [review]
There are no special considerations necessary. Although I would test a single dag locally to ensure that everything is working before merging. The clean-gke-pods dag is an easy one to use.

this may help: https://mana.mozilla.org/wiki/pages/viewpage.action?spaceKey=DOPS&title=WTMO+Developer+Guide#WTMODeveloperGuide-LocalTesting&DevelopmentofDAGs

**Comment by haroldwoo** [comment]
I would also note that 

1. We should avoid extending the DAG object or any other airflow primitive when possible. This potentially becomes a source of headaches when upgrading.
2. Jeff's mention of adding tag validation to CI is a great idea but can be done in a follow up PR if need be.

File: requirements.in
Comment (kik-kik) [reply requirements.in:22]: yes, that is a good point. To be honest I was a bit unsure whether I should add pytest or not. Could there be a particular reason why we don't currently have it? Perhaps, @jklukas would you know?

**Comment by kik-kik** [comment]
> I would also note that
> 
>     1. We should avoid extending the DAG object or any other airflow primitive when possible. This potentially becomes a source of headaches when upgrading.
> 
>     2. Jeff's mention of adding tag validation to CI is a great idea but can be done in a follow up PR if need be.

1. I can see that.
2. I agree, I think that would be quite nice to have. I'll create a new bug or issue (not sure if it matters) and assign it to myself.

Thanks a lot for your input @haroldwoo 🙇

File: requirements.in
Comment (jklukas) [reply requirements.in:22]: To my knowledge, this repo has not historically had any unit tests, meaning there was no need for a pytest dependency. The primary case that we added testing for was whether dags are parseable at all, for which we built the bash script that lists DAGs, thus causing the python code to be loaded and exposing any syntax errors.

**Comment by kik-kik** [comment]
Hey @haroldwoo,

I followed the instructions linked, I noticed that when trying to start Airflow locally with `make up` mysql image failed to be pulled with the following error:

 I did a quick Google and was able to find a fix/get it working: 

Would it be okay for me to add `platform: linux/x86_64` to the docker-compose file for db container definition? It would like this:

```yaml
[...]
services:
  db:
    platform: linux/x86_64
    image: mysql:5.7
    ports:
      - '3306:3306'
[...]
```

**Comment by kik-kik** [comment]
Started Airflow locally and it appears tags are showing up:
image

Gonna try running 1 DAG but since I don't see any errors I don't believe this change should cause any problems.

**Comment by kik-kik** [comment]
I created a DAG locally to just run bash commands locally and it appears Airflow worked correctly:
image

Is that good enough or is more testing needed? @haroldwoo

**Review by haroldwoo — APPROVED** [review]

**Comment by haroldwoo** [comment]
The CI is failing because any .py file under the dags dir will be treated as a dag itself.

[2022-01-06 15:59:50,255] {{dagbag.py:329}} ERROR - Failed to import: /app/dags/tests/utils/test_tags.py
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airflow/models/dagbag.py", line 326, in _load_modules_from_file
    loader.exec_module(new_module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/app/dags/tests/utils/test_tags.py", line 2, in <module>
    from dags.utils.tags import Tag, InvalidTagError
ModuleNotFoundError: No module named 'dags'

Move your unit tests outside the dags/ dir.

As for the mysql portion, what is your error? It may be local to your environment since others are able to start mysql. Have you gone through all the steps in the README (including increasing the memory limit for Docker desktop)?

**Comment by kik-kik** [comment]
Hey @haroldwoo ,

As I mentioned, I believe this error is due to me using M1 Macbook. I included the error in my last comment:

> I followed the instructions linked, I noticed that when trying to start Airflow locally with `make up` mysql image failed to be pulled with the following error: `no matching manifest for linux/arm64/v8 in the manifest list entries`
> 
> I did a quick Google and was able to find a fix/get it working: https://stackoverflow.com/questions/65456814/docker-apple-silicon-m1-preview-mysql-no-matching-manifest-for-linux-arm64-v8
> 
> Would it be okay for me to add `platform: linux/x86_64` to the docker-compose file for db container definition? It would like this:
> 
> ```yaml
> [...]
> services:
>   db:
>     platform: linux/x86_64
>     image: mysql:5.7
>     ports:
>       - '3306:3306'
> [...]
> ```
> 
> Seems to be mac M1 specific thing.

`no matching manifest for linux/arm64/v8 in the manifest list entries` - the stackoverflow discussion helped me fix it.

**Comment by haroldwoo** [comment]
Adding the line: platform: linux/x86_64

Causes my env to break: ERROR: The Compose file './docker-compose.yml' is invalid because:
Unsupported config option for services.db: 'platform'

You may be able to figure out how to add a conditional that checks local environment and acts accordingly

**Comment by kik-kik** [comment]
@haroldwoo hmm, ok. I'll take a note of this and explore it further. 

@jklukas out of curiosity could you please try and see if you see the same error as me when trying to start airflow locally? If it indeed is just my machine then it's probably not worth investing time into this.

**Comment by haroldwoo** [comment]
nevermind, I updated docker-compose and it seems ok

**Comment by kik-kik** [comment]
> nevermind, I updated docker-compose and it seems ok

ok, cool. Thanks!

**Comment by haroldwoo** [comment]
Just a nit, but please squash your commits before merging in the future. It keeps the log cleaner

**Comment by kik-kik** [comment]
Hmm, I thought I pressed squad and merge 🤔

**Pull request merged**
